### PR TITLE
Fix synchronization in debug mode cupy injection test

### PR DESCRIPTION
**Pull request by ksztenderski**
Signed-off-by: ksztenderski <ksztenderski@nvidia.com>

## Category:

**Bug fix** (*non-breaking change which fixes an issue*)

## Description:

Fixes synchronization issue with cupy injection test of debug pipeline.
Test failure example: [https://gitlab-master.nvidia.com/dl/dali/dali-gh-mirror/-/jobs/39133275](https://gitlab-master.nvidia.com/dl/dali/dali-gh-mirror/-/jobs/39133275)

### Affected modules and functionalities:

Changed `test_injection_cupy` implementation.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by ksztenderski** [comment]
!build

**Comment by klecki** [comment]
I am thinking it this is indeed the problem of the test, or if it is actually more general problem in the debug mode. I think, we have some code around the feed_input to extract the "best guess" stream to use when copying the user-provided data using `_get_default_stream_for_array` or to allow user to overwrite the stream (see `cuda_stream` parameter to external source).

I think we should probably replicate that behaviour, infer the stream and if necessary in debug mode, add a sync between the stream of the input data and the internal stream we are using in debug mode. This fix probably just hides the problem.

File: dali/test/python/test_pipeline_debug.py [hunk 1]
```diff
@@ -153,7 +153,13 @@ def test_injection_torch():
 @attr('cupy')
 def test_injection_cupy():
     import cupy
-    _test_injection('gpu', 'cupy array', lambda xs: [cupy.array(x) for x in xs])
+
+    def get_data(xs):
+        res = [cupy.array(x) for x in xs]
+        cupy.cuda.Device().synchronize()
+        return res
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/test/python/test_pipeline_debug.py:160]: I'd say this is disturbing. Why is it necessary? I think it may be a symptom of a larger problem.

**Comment by ksztenderski** [comment]
Closing, needs fix for synchronization with inputs from outer sources (torch, cupy), both in debug mode `external_source` and injections.

**Pull request closed without merging**